daemon/containerd/image_push.go

```diff
@@ -80,12 +87,219 @@ func (i *ImageService) PushImage(ctx context.Context, image, tag string, metaHea
 	finishProgress := showProgress(ctx, jobs, outStream, pushProgress(tracker))
 	defer finishProgress()
 
-	logrus.WithField("desc", target).WithField("ref", ref.String()).Info("Pushing desc to remote ref")
-	err = i.client.Push(ctx, ref.String(), target,
-		containerd.WithResolver(resolver),
-		containerd.WithPlatformMatcher(platformMatcher),
-		containerd.WithImageHandler(imageHandler),
-	)
+	return push(ctx, store, ref.String(), target, resolver, imageHandler)
+}
+
+// Push uploads the provided content to a remote resource
+func push(ctx context.Context, store content.Store, ref string, desc ocispec.Descriptor, resolver remotes.Resolver, imagesHandler containerdimages.HandlerFunc) error {
+	// Annotate ref with digest to push only push tag for single digest
+	if !strings.Contains(ref, "@") {
+		ref = ref + "@" + desc.Digest.String()
+	}
+
+	pusher, err := resolver.Pusher(ctx, ref)
+	if err != nil {
+		return err
+	}
+
+	wrapper := func(h images.Handler) images.Handler {
+		h = images.Handlers(imagesHandler, h)
+
+		return h
+	}
+
+	sources, err := collectSources(ctx, desc, store)
+	if err != nil {
+		return err
+	}
+
+	lazyStore := newLazyContentStore(store, sources)
+
+	var limiter *semaphore.Weighted
+	return remotes.PushContent(ctx, pusher, desc, lazyStore, limiter, platforms.All, wrapper)
+}
+
+func findLazyChildren(ctx context.Context, desc ocispec.Descriptor, store content.Store) ([]ocispec.Descriptor, error) {
+	// Collect to hashset to remove duplicates
+	set := map[string]ocispec.Descriptor{}
+
+	// Do a breadth-first search starting from this descriptor
+	queue := []ocispec.Descriptor{desc}
+	for len(queue) > 0 {
+		child := queue[0]
+		queue = queue[1:]
+
+		if containerdimages.IsNonDistributable(child.MediaType) {
+			continue
+		}
+
+		_, err := store.ReaderAt(ctx, child)
+		if err != nil && cerrdefs.IsNotFound(err) {
+			set[child.Digest.String()] = child
+			continue
+		}
+
+		newChildren, err := containerdimages.Children(ctx, store, child)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(newChildren) > 0 {
+			queue = append(queue, newChildren...)
+		}
+	}
+
+	result := []ocispec.Descriptor{}
+	for _, desc := range set {
+		result = append(result, desc)
+		logrus.WithField("desc", desc).Debug("lazy children found")
+	}
+
+	return result, nil
+
+}
+
+// Do a small peek of the content to check if content is definitely not a json.
+// Returns true when content is definitely not a json.
+// If it returns false, then the content may still not be a json.
```

Comment:
```suggestion
// peekNotJson  does a small peek of the content to check if content is definitely not JSON.
// It returns true if content is definitely not JSON, or false if it was unable to detect if it's
// JSON or not.
```

The error doesn't not appear to be described (on how it should be used in combination with the bool)

Reply:
Thanks. I didn't bother much describing the error case, as it's not really something that should happen under normal circumstances. This is called only when traversing the content store, so we expect the content to be there.
The only case of error here would be if the content is silently deleted in background between the beginning of the Walk and running handler for this particular content.
BTW, I wonder if there is some way to "lock" the store, so we can be 100% sure that this won't happen.

Reply:
I think you can use a leased context maybe?

Reply:
But how do I put a lease on everything in the Content store? :P
And if I do - will lease be enough to pause something like `ctr content remove <hash>` until the Walk is finished?

Reply:
Yeah nah it doesn't work, the lease you add to your context is added to objects you create during that context.